### PR TITLE
Add Annotation Inheritance, tests, and docs

### PR DESCRIPTION
**Pull request by briandowns**
Resolves: #633

### What's Changed:

Added the ability to have classes inherit class, method, and field annotations.

#

### Type of Change:

- [ ] Bug fix
- [x] New feature
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

#

### Housekeeping:

- [ ] Tests have been updated to reflect the changes done within this PR (if applicable).
- [x] Documentation has been updated to reflect the changes done within this PR (if applicable).

#

### CODE REVIEW

File: src/vm/vm.c [hunk 1]
```diff
@@ -796,6 +811,20 @@ static void setReplVar(DictuVM *vm, Value value) {
...
+    for (int i = 0; i <= superAnnotations->capacityMask; ++i) {
+        if (IS_EMPTY(superAnnotations->entries[i].key)) {
+            continue;
+        }
+
+        Value val = superAnnotations->entries[i].value;
+        push(vm, val);
+
+        dictSet(vm, klassAnnotations, superAnnotations->entries[i].key, val);
+        pop(vm);
```
Comment (Jason2605) [inline src/vm/vm.c:824]: We shouldn't need to push to the stack here as they should already be referenced in the parent class, which means they will get greyed there

File: src/vm/vm.c [hunk 2]
```diff
@@ -2187,6 +2216,20 @@ static DictuInterpretResult run(DictuVM *vm) {
             ObjDict *dict = AS_DICT(READ_CONSTANT());
             ObjClass *klass = AS_CLASS(peek(vm, 0));
 
+            if (dict->entries == 0) {
+                printf("XXX - 0!\n");
+            }
```
Comment (Jason2605) [inline src/vm/vm.c:2221]:
```suggestion
```

**Comment by Jason2605** [comment]
Couple of things we will need to think about with this when testing actually - the first is if the child class overrides something in the parent class we want to take the childs implementation not the parents

```
class Base {
    @method
    test() {

    }
}

class Test < Base {
    @method1234
    test() {

    }
}

print(Test.methodAnnotations); // {"test": {"method": nil}}

// Where I would expect {"test": {"method1234": nil}}
```

I would imagine some sort of key check in copyAnnotations should do the trick.

Similarly (not sure why this would happen) but if we modify the annotation value on the child we want to make sure it doesn't update the parent (essentially we would want a deep copy)

```
class Base {
    @method
    test() {

    }
}

class Test < Base {

}

Test.methodAnnotations['test']['method'] = 10;

print(Test.methodAnnotations); // {"test": {"method": 10}}
print(Base.methodAnnotations); // {"test": {"method": 10}}
```

**Comment by briandowns** [comment]
Great catches. Impl updated as well as tests.

File: src/vm/vm.c [hunk 3]
```diff
@@ -761,6 +761,21 @@ static void createClass(DictuVM *vm, ObjString *name, ObjClass *superclass, Clas
     if (superclass != NULL) {
         tableAddAll(vm, &superclass->publicMethods, &klass->publicMethods);
         tableAddAll(vm, &superclass->abstractMethods, &klass->abstractMethods);
+
+        if (superclass->classAnnotations != NULL) {
+            ObjDict *dict = copyDict(vm, superclass->classAnnotations, true);
```
Comment (Jason2605) [inline src/vm/vm.c:766]:
```suggestion
            ObjDict *dict = copyDict(vm, superclass->classAnnotations, false);
```

File: src/vm/vm.c [hunk 4]
```diff
@@ -761,6 +761,21 @@ static void createClass(DictuVM *vm, ObjString *name, ObjClass *superclass, Clas
...
+            klass->classAnnotations = dict;
+        }
+
+        if (superclass->methodAnnotations != NULL) {
+            ObjDict *dict = copyDict(vm, superclass->methodAnnotations, true);
```
Comment (Jason2605) [inline src/vm/vm.c:771]:
```suggestion
            ObjDict *dict = copyDict(vm, superclass->methodAnnotations, false);
```

File: src/vm/vm.c [hunk 5]
```diff
@@ -761,6 +761,21 @@ static void createClass(DictuVM *vm, ObjString *name, ObjClass *superclass, Clas
...
+            klass->methodAnnotations = dict;
+        }
+
+        if (superclass->fieldAnnotations != NULL) {
+            ObjDict *dict = copyDict(vm, superclass->fieldAnnotations, true);
```
Comment (Jason2605) [inline src/vm/vm.c:776]:
```suggestion
            ObjDict *dict = copyDict(vm, superclass->fieldAnnotations, false);
```

File: src/vm/vm.c [hunk 6]
```diff
@@ -2187,6 +2219,16 @@ static DictuInterpretResult run(DictuVM *vm) {
...
 
+            if (IS_CLASS(peek(vm, 1))) {
+                ObjClass *super = AS_CLASS(peek(vm, 1));
+
+                if (super->classAnnotations != NULL) {
+                    ObjDict *superAnnotations = copyDict(vm, super->classAnnotations, true);
+
+                    copyAnnotations(vm, superAnnotations, dict);
+                }
+            }
```
Comment (Jason2605) [inline src/vm/vm.c:2230]: I just kinda had a realisation with this bit of code as well (and the other two opcodes)

We can simply just do:
```suggestion
            if (klass->classAnnotations != NULL) {
                copyAnnotations(vm, klass->classAnnotations, dict);
            }
```

As the dict on the class has already been copied from the parent so we should be free to modify this directly

**Review by Jason2605 — COMMENTED** [review]
Sorry couple more comments!

**Comment by briandowns** [comment]
All set!

**Comment by Jason2605** [comment]
Nicceeee thanks for this!!

**Comment by briandowns** [comment]
No prob. Thanks for the quick reviews! My ORM is really coming along now that all of the annotation features are in!

**Comment by Jason2605** [comment]
Awesome! Definitely be interested to see it! Also side note, I ended up using Wanbli for a little side project of mine :)

**Comment by briandowns** [comment]
Open to any and all feedback on that. I don't really have it in a stable state yet so use with caution!

**Comment by Jason2605** [comment]
It’s all good just a little tool I’m using in the house to track some smart plug usage! I’ll potentially PR some things across for it but it’s been a fun project

**Comment by briandowns** [comment]
Here's the ORM. Suuuuper naive... Not finished but works. https://github.com/briandowns/kahless . Gonna get an example together using Kahless and Wanbli together.

**Pull request merged**